File: frontend/src/types/location.ts
export interface ApiLocation {
  id: number;
  room: string;
  shelf: string;
  layer: string;
  position: string;
  full_path: string;
  description: string;
  sort_order: number;
  created_at?: string;
  updated_at?: string;
}

export interface Location {
  id: number;
  room: string;
  shelf: string;
  layer: string;
  position: string;
  fullPath: string;
  description: string;
  sortOrder: number;
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateLocationPayload {
  room: string;
  shelf: string;
  layer: string;
  position: string;
  description?: string;
  sort_order?: number;
}

export interface UpdateLocationPayload {
  room?: string;
  shelf?: string;
  layer?: string;
  position?: string;
  description?: string;
  sort_order?: number;
}

export interface LocationBrief {
  id: number;
  fullPath: string;
}

export function normalizeLocation(api: ApiLocation): Location {
  return {
    id: api.id,
    room: api.room,
    shelf: api.shelf,
    layer: api.layer,
    position: api.position,
    fullPath: api.full_path,
    description: api.description,
    sortOrder: api.sort_order,
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

File: backend/app/schemas/location.py
from __future__ import annotations
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, model_validator


class LocationBase(BaseModel):
    room: str
    shelf: str
    layer: Optional[str] = None
    position: Optional[str] = None
    description: Optional[str] = None
    sort_order: int = 0


class LocationCreate(LocationBase):
    pass


class LocationUpdate(BaseModel):
    room: Optional[str] = None
    shelf: Optional[str] = None
    layer: Optional[str] = None
    position: Optional[str] = None
    description: Optional[str] = None
    sort_order: Optional[int] = None


class LocationResponse(LocationBase):
    id: int
    full_path: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LocationBrief(BaseModel):
    id: int
    full_path: str

    model_config = {"from_attributes": True}
